sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,1061 @@
+    default:
+      return kDifUsbdevTxPacketWriteBytesResultError;
+  }
+
+  size_t remaining_buffer_space = USBDEV_BUFFER_ENTRY_SIZE_BYTES - buffer_len;
+  // `bytes_to_copy` is the minimum of `src_len` and `remaining_buffer_space`.
+  size_t bytes_to_copy = src_len;
+  if (bytes_to_copy > remaining_buffer_space) {
+    bytes_to_copy = remaining_buffer_space;
+  }
+
+  // Write bytes to the buffer
+  const uint32_t buffer_addr = get_buffer_addr(buffer_entry, buffer_len);
+  mmio_region_memcpy_to_mmio32(usbdev->base_addr, buffer_addr, src,
+                               bytes_to_copy);
+
+  buffer_len += bytes_to_copy;
+  remaining_buffer_space -= bytes_to_copy;
+
+  // Update buffer length in the USBDEV_CONFIGINX register.
+  mmio_region_nonatomic_set_field32(
+      usbdev->base_addr, config_in_reg_offset,
+      (bitfield_field32_t){.mask = USBDEV_CONFIGIN0_SIZE0_MASK,
+                           .index = USBDEV_CONFIGIN0_SIZE0_OFFSET,
+                           .value = buffer_len});
+
+  if (bytes_written) {
+    *bytes_written = bytes_to_copy;
+  }
+
+  if (remaining_buffer_space == 0 && bytes_to_copy < src_len) {
+    return kDifUsbdevTxPacketWriteBytesResultBufferFull;
+  }
+
+  return kDifUsbdevTxPacketWriteBytesResultOK;
+}
+
+dif_usbdev_tx_packet_send_result_t dif_usbdev_tx_packet_send(
+    dif_usbdev_t *usbdev, uint8_t endpoint) {
+  if (usbdev == NULL || !is_valid_endpoint(endpoint)) {
+    return kDifUsbdevTxPacketSendResultBadArg;
+  }
+
+  // Make sure that there is an active write operation for the endpoint
+  if (!buffer_op_is_writing(usbdev, endpoint)) {
+    return kDifUsbdevTxPacketSendResultNotWriting;
+  }
+
+  // Stop writing to buffer for this endpoint
+  buffer_op_stop_write(usbdev, endpoint);
+  // Mark the packet as ready for transmission
+  mmio_region_nonatomic_set_bit32(
+      usbdev->base_addr, kEndpointHwInfos[endpoint].config_in_reg_offset,
+      USBDEV_CONFIGIN0_RDY0);
+
+  return kDifUsbdevTxPacketSendResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_tx_packet_get_status(
+    dif_usbdev_t *usbdev, uint8_t endpoint,
+    dif_usbdev_tx_packet_status_t *status) {
+  if (usbdev == NULL || status == NULL || !is_valid_endpoint(endpoint)) {
+    return kDifUsbdevBadArg;
+  }
+
+  // Make sure that there is no active write buffer operation for the endpoint
+  if (buffer_op_is_writing(usbdev, endpoint)) {
+    *status = kDifUsbdevTxPacketStatusStillWriting;
+    return kDifUsbdevOK;
+  }
+
+  // HW details
+  const uint32_t config_in_reg_offset =
+      kEndpointHwInfos[endpoint].config_in_reg_offset;
+  const uint8_t endpoint_bit_index = kEndpointHwInfos[endpoint].bit_index;
+
+  // Buffer used by this endpoint
+  const uint8_t buffer = mmio_region_read_mask32(
+      usbdev->base_addr, config_in_reg_offset, USBDEV_CONFIGIN0_BUFFER0_MASK,
+      USBDEV_CONFIGIN0_BUFFER0_OFFSET);
+
+  // Check the status of the packet
+  if (mmio_region_get_bit32(usbdev->base_addr, config_in_reg_offset,
```

Comment:
Would it make sense to reorganize this control flow to avoid re-reading this particular register multiple times?

Reply:
Done.